workflow-runtime/src/main/java/com/squareup/workflow1/WorkflowInterceptor.kt

```diff
@@ -165,3 +264,55 @@ internal fun <P, S, O, R> WorkflowInterceptor.intercept(
     override fun toString(): String = "InterceptedWorkflow($workflow, $this@intercept)"
   }
 }
+
+@OptIn(ExperimentalWorkflowApi::class)
+private class InterceptedRenderContext<P, S, O>(
+  private val baseRenderContext: BaseRenderContext<P, S, O>,
+  private val interceptor: RenderContextInterceptor<P, S, O>
+) : BaseRenderContext<P, S, O>, Sink<WorkflowAction<P, S, O>> {
+  override val actionSink: Sink<WorkflowAction<P, S, O>> get() = this
+
+  override fun send(value: WorkflowAction<P, S, O>) {
+    interceptor.onActionSent(value) { interceptedAction ->
+      baseRenderContext.actionSink.send(interceptedAction)
+    }
+  }
+
+  override fun <ChildPropsT, ChildOutputT, ChildRenderingT> renderChild(
+    child: Workflow<ChildPropsT, ChildOutputT, ChildRenderingT>,
+    props: ChildPropsT,
+    key: String,
+    handler: (ChildOutputT) -> WorkflowAction<P, S, O>
+  ): ChildRenderingT =
+    interceptor.onRenderChild(child, props, key, handler) { c, p, k, h ->
+      // TODO better lambda arg names
+      baseRenderContext.renderChild(c, p, k, h)
+    }
+
+  override fun runningSideEffect(
+    key: String,
+    sideEffect: suspend CoroutineScope.() -> Unit
+  ) {
+    baseRenderContext.runningSideEffect(key) {
+      val outerScope = this
+
+      // Only invoke the hook when the side effect is actually starting – i.e. don't invoke
+      // it on subsequent consecutive calls with the same key.
+      interceptor.onSideEffectRunning(key) {
+        // We don't want to use the CoroutineScope receiver from the outer runningSideEffect
+        // here since the interceptor could be calling this function from a different context
+        // (e.g. wrap it with a different dispatcher or name).
+        coroutineScope {
+          val innerScope = this
+          // If the interceptor is wrapping the coroutine context, it can't change the job
+          // because that could break structured concurrency.
+          check(outerScope.coroutineContext.job === innerScope.coroutineContext.job) {
+            // TODO unit tests
+            "Expected onSideEffectStarting not to call proceed with a different Job."
+          }
+          sideEffect()
+        }
```

Comment:
```suggestion
      interceptor.onSideEffectRunning(key) {
        // We don't want to use the CoroutineScope receiver from the outer runningSideEffect
        // here since the interceptor could be calling this function from a different context
        // (e.g. wrap it with a different dispatcher or name).
        // TODO(ray): Double-check that `coroutineContext` here is the `suspend val` property, and not the property from the outer scope receiver.
        val innerScope = CoroutineScope(coroutineContext)
        // If the interceptor is wrapping the coroutine context, it can't change the job
        // because that could break structured concurrency.
        check(outerScope.coroutineContext.job === innerScope.coroutineContext.job) {
          // TODO unit tests
          "Expected onSideEffectStarting not to call proceed with a different Job."
        }
        innerScope.sideEffect()
      }
```

Reply:
Remember when we believed that `withContext` just passes the `Job` through? Those were simpler times.

```kotlin
            println("1. ${coroutineContext.job}")
            withContext(TestElement) {
              println("2. ${coroutineContext.job}")
              proceed()
            }
```

```
1. "coroutine#3":BlockingCoroutine{Active}@77847718
2. "coroutine#3":UndispatchedCoroutine{Active}@9597028
```

Reply:
This does the trick:

```kotlin
            println("1. ${activeCoroutineContext().job}")
            CoroutineScope(TestElement).run {
              println("2. ${activeCoroutineContext().job}")
              proceed
            }
```

Reply:
Also, this doesn't do what you think it does. The scope receiver has no effect on the `CoroutineContext` that `sideEffect` runs in.

```kotlin
        val innerScope = CoroutineScope(coroutineContext)
        innerScope.sideEffect()
```

But this does:

```kotlin
        val innerScope = CoroutineScope(coroutineContext)
        innerScope.launch { sideEffect() }
```